### PR TITLE
Improve compatibility with non-whole-minute timezone offsets

### LINKED ISSUE #27079: Incorrect responses with timezone offsets that are not whole hours
You have to provide the following information whenever possible.

**Describe the bug**

Appeared at https://github.com/ClickHouse/ClickHouse/pull/26978 CI (random timezone)

If the timezone offset isn't a whole hour, like Africa/Monrovia which used to be -2670 (` -0:44:30`), several functions like `formatDateTime` or `toStartOfMinute` won't work correctly.

**Does it reproduce on recent release?**

Yes, and probably all stable releases are affected.

**How to reproduce**

```sql
WITH toDateTime('1970-06-17 07:39:21', 'Africa/Monrovia') AS t
SELECT
    toUnixTimestamp(t),
    timeZoneOffset(t),
    formatDateTime(t, '%F %T', 'Africa/Monrovia'),
    toString(t, 'Africa/Monrovia'),
    toStartOfMinute(t),
    toStartOfFiveMinute(t),
    toStartOfFifteenMinutes(t),
    toStartOfTenMinutes(t),
    toStartOfHour(t)
FORMAT Vertical

Query id: 8c580961-9679-406f-8555-5d0ad5a29d36

Row 1:
──────
toUnixTimestamp(t):                            14459031
timeZoneOffset(t):                             -2670
formatDateTime(t, '%F %T', 'Africa/Monrovia'): 1970-06-17 07:39:51
toString(t, 'Africa/Monrovia'):                1970-06-17 07:39:21
toStartOfMinute(t):                            1970-06-17 07:38:30
toStartOfFiveMinute(t):                        1970-06-17 07:35:30
toStartOfFifteenMinutes(t):                    1970-06-17 07:30:30
toStartOfTenMinutes(t):                        1970-06-17 07:30:00
toStartOfHour(t):                              1970-06-17 07:00:00

1 rows in set. Elapsed: 0.002 sec. 

```

**Expected behavior**

```
toUnixTimestamp(t):                            14459031
timeZoneOffset(t):                             -2670
formatDateTime(t, '%F %T', 'Africa/Monrovia'): 1970-06-17 07:39:21
toString(t, 'Africa/Monrovia'):                1970-06-17 07:39:21
toStartOfMinute(t):                            1970-06-17 07:39:00
toStartOfFiveMinute(t):                        1970-06-17 07:35:00
toStartOfFifteenMinutes(t):                    1970-06-17 07:30:00
toStartOfTenMinutes(t):                        1970-06-17 07:30:00
toStartOfHour(t):                              1970-06-17 07:00:00
```

It seems that DateLut needs to take into account that offsets can change not only full hours, but minutes and even seconds. I'll create a PR with some suggested changes to improve the situation.

### PR DESCRIPTION
**Pull request by Algunenano**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve compatibility with non-whole-minute timezone offsets

Detailed description / Documentation draft:
DateLUT was assuming timezone offsets are whole minutes but that's not true, for example Liberia used to be `-0:43:08` and later became `-0:44:30` (it currently is 00:00, but that's less exciting).

Closes https://github.com/ClickHouse/ClickHouse/issues/27079
It appeared as part of the CI (in https://github.com/ClickHouse/ClickHouse/pull/26978)

### CODE REVIEW

**Comment by Algunenano** [comment]
2 things in CI:
* Multiple checksum reports (https://clickhouse-test-reports.s3.yandex.net/27080/f2f778376295c077224fcd6f9242f6f0164a4e5f/functional_stateful_tests_(address).html#fail1): it looks like a hardware issue as I've built ASAN locally and run those tests without issues. The odd thing is that they are connection to localhost, maybe memory?
* PVS check. I'll try to fix it

Let's see if a new commit fixing (2) also shows the address issues or it has been something temporary.

**Comment by amosbird** [comment]
> for example Liberia used to be `-0:43:08`

That's interesting...

**Comment by Algunenano** [comment]
Rebased with master. No new changes (it was clean already)

**Comment by alexey-milovidov** [comment]
We have 2x performance degradation of `toSecond` function or somewhat similar:
https://clickhouse-test-reports.s3.yandex.net/27080/1b58e276b8b0534e925c79e9764896c847f9ec7f/performance_comparison/report.html#fail1

File: base/common/DateLUTImpl.h [hunk 1]
```diff
@@ -474,38 +493,23 @@ class DateLUTImpl
...
         LUTIndex index = findIndex(t);
-        UInt32 time = t - lut[index].date;
+        Time time = t - lut[index].date;
 
         if (time >= lut[index].time_at_offset_change())
             time += lut[index].amount_of_offset_change();
 
-        return time / 60 % 60;
+        Time res = time / 60 % 60;
+        if (likely(res >= 0))
```
Comment (alexey-milovidov) [inline base/common/DateLUTImpl.h:502]: How the opposite is possible if we are dividing the number of seconds after midnight (always positive)?

File: base/common/DateLUTImpl.h [hunk 2]
```diff
@@ -459,7 +464,21 @@ class DateLUTImpl
 
     inline unsigned toSecond(Time t) const
     {
-        auto res = t % 60;
+        if (offset_is_whole_number_of_hours_during_epoch)
```
Comment (alexey-milovidov) [inline base/common/DateLUTImpl.h:467]: likely?

**Comment by Algunenano** [comment]
> We have 2x performance degradation of toSecond function or somewhat similar:

It might be because `Asia/Kolkata` is at `UTC+5:30` and that won't use the fast path (`offset_is_whole_number_of_hours_during_epoch`). I'll see if there is something better to do in this situation,

**Comment by alexey-milovidov** [comment]
I'm not sure about this change.
We did not support non whole 15 minutes offsets - this is deliberate choice.

... but if we can avoid any performance degradation.

File: base/common/DateLUTImpl.h (hunk 1 above)
Comment (Algunenano) [reply base/common/DateLUTImpl.h:502]: I was probably thinking about negative values of `t` (with Datetime64), which will lead to negative `time`  and negative `res`; but that doesn't see to be supported anyway.

**Comment by Algunenano** [comment]
I think the performance degradation can be solved in a similar way as the whole hour thing, by adding another flag like `offset_is_whole_number_of_minutes_during_epoch`. It's not great, but it should limit the performance impact to the offsets were the responses were wrong. I'll give it a try (likely on Monday).

**Comment by alexey-milovidov** [comment]
Ok, sounds good.

**Comment by Algunenano** [comment]
Rebased and applied the improvements from the comments.

Performance of the affected query:
* Initially (before changes): `QPS: 12.465, RPS: 623806343.190, MiB/s: 4759.265, result RPS: 12.465`
* After initial changes: `QPS: 3.516, RPS: 175981641.031, MiB/s: 1342.633, result RPS: 3.516`
* Now: `QPS: 12.487, RPS: 624930109.263, MiB/s: 4767.838, result RPS: 12.487`

Looking at how things work internally (DateTimeTransforms.h), it looks like the fast path (common offsets) of some of the functions (toSecond, toMinute) could be vectorized using the same architecture as other functions (accepting an array instead of a single element), but that's something for a different time.

**Comment by alexey-milovidov** [comment]
Ok.

Theoretically it can also be auto-vectorized, because the function is inlined. But I doubt that compiler is smart enough to perform "unlikely" branch checks in batches...

**Review by alexey-milovidov — APPROVED** [review]

**Comment by Algunenano** [comment]
I added an invalid optimization (ok for minutes, not for hours). I'm looking into it and will try to unify calls/code if possible. Sadly, changing DateLutImpl.h requires recompiling ~1300 files so it takes time to verify and benchmark

**Comment by alexey-milovidov** [comment]
> GithubException: 500 null

Looks like GitHub API is down.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
@Mergifyio update

**Review by alexey-milovidov — APPROVED** [review]
Wow, everything is alright :)

**Pull request merged**